File: dynamic_programming/746_min_cost_climbing_stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        # n = len(cost)
        # dp = [0] * n

        # for i in range(n):
        #     if i < 2:
        #         dp[i] = cost[i]
        #     else:
        #         dp[i] = cost[i] + min(dp[i - 1], dp[i - 2])

        # return min(dp[-1], dp[-2]) 
        n = len(cost)
        dp = [0] * (n + 1)

        for i in range(2, n + 1):
            dp[i] = min(dp[i-1] + cost[i - 1], dp[i - 2] + cost[i - 2])

        return dp[n]
             
      

File: dynamic_programming/746_min_cost_climbing_stairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int[] dp = new int[n + 1];
        
        for (int i = 2; i <= n; i++) {
            int oneStepCost = dp[i - 1] + cost[i - 1];
            int twoStepCost = dp[i - 2] + cost[i - 2];
            dp[i] = Math.min(oneStepCost, twoStepCost);
        }
        return dp[n];
    }
}